src/GraphQL.MemoryCache/AutomaticPersistedQueries/AutomaticPersistedQueriesExecution.cs

```diff
@@ -0,0 +1,120 @@
+using System.Security.Cryptography;
+using System.Text;
+using GraphQL.DI;
+
+namespace GraphQL.Caching;
+
+/// <summary>
+/// Implementation of Automatic Persisted Queries (APQ).
+/// https://www.apollographql.com/docs/react/api/link/persisted-queries/
+/// </summary>
+public class AutomaticPersistedQueriesExecution : IConfigureExecution
+{
+    private readonly IQueryCache _cache;
+    private const string SUPPORTED_VERSION = "1";
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public AutomaticPersistedQueriesExecution(IQueryCache cache)
+    {
+        _cache = cache;
+    }
+
+    /// <summary>
+    /// Searching APQ fields in <see cref="ExecutionOptions.Extensions"/> based on a protocol:
+    /// https://www.apollographql.com/docs/react/api/link/persisted-queries/#protocol
+    /// </summary>
+    protected virtual (string? Hash, string? Version, bool Enabled) GetAPQProperties(Inputs? extensions)
+    {
+        string? hashResult = null;
+        string? versionResult = null;
+        bool enabled = false;
+
+        if (extensions?.TryGetValue("persistedQuery", out var persistedQueryObject) ?? false)
+        {
+            enabled = true;
+
+            if (persistedQueryObject is Dictionary<string, object> persistedQuery)
+            {
+                if (persistedQuery.TryGetValue("sha256Hash", out var sha256HashObject) && sha256HashObject is string sha256Hash && !string.IsNullOrWhiteSpace(sha256Hash))
+                {
+                    hashResult = sha256Hash;
+                }
+
+                if (persistedQuery.TryGetValue("version", out var versionObject) && versionObject is string version && !string.IsNullOrWhiteSpace(version))
+                {
+                    versionResult = version;
+                }
+            }
+        }
+
+        return (hashResult, versionResult, enabled);
+    }
+
+    /// <summary>
+    /// Compute SHA256 hash.
+    /// </summary>
+    protected virtual string ComputeSHA256(string input)
+    {
+        var bytes = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(input));
```

Comment:
`SHA256.Create()` may be moved into static.

Reply:
And a general note regarding this method - it's greedy for memory.

Reply:
Are you saying that you can use a static instance of SHA256?  I do not see in the documentation that instance members can be called from multiple threads in parallel.  Do you believe that is the case?  If so, why?

I would expect that we can write some optimized code that reduces memory usage here.  I would probably write such an implementation in a separate PR if we choose to do so.

Reply:
> Are you saying that you can use a static instance of SHA256?

I believe yes but in general this method should be called rather rarely and bytes <-> string conversion takes much more memory.

Reply:
Confirmed it cannot be static.


```cs
    [Fact]
    public async Task sha256test()
    {
        var sha256shared = SHA256.Create();
        var tasks = new List<Task>();

        // spin up 8 concurrent threads
        for (int i = 0; i < 8; i++)
        {
            tasks.Add(Task.Factory.StartNew(() =>
            {
                var rng = RandomNumberGenerator.Create();
                var b = new byte[4];
                rng.GetBytes(b);
                var r = new Random(BitConverter.ToInt32(b, 0));
                for (int j = 0; j < 100; j++)
                {
                    b = new byte[r.Next(8192, 10240)];
                    rng.GetBytes(b);
                    var hash1 = SHA256.Create().ComputeHash(b);

                    // passes
                    var hash2 = sha.ComputeHash(b);
                    hash1.ShouldBe(hash2, customMessage: "new instance different than local instance");

                    // fails
                    var hash3 = sha256shared.ComputeHash(b);
                    hash1.ShouldBe(hash3, customMessage: "new instance different than shared instance");
                }
            }));
        }
        foreach (var t in tasks)
        {
            await t.ConfigureAwait(false);
        }
    }
```

Reply:
>  var hash2 = sha.ComputeHash(b);

Where is `sha` variable?

Reply:
Supposed to be right after “rng= “